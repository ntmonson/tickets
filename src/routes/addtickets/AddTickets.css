@import '../../components/variables.css';

.root {
  padding-left: 20px;
  padding-right: 20px;
}

.container {
  margin: 0 auto;
  padding: 0 0 40px;
  max-width: var(--screen-lg-min);
}

textarea {
  width: 100%;
  height: 150px;
  padding: 12px 20px;
  box-sizing: border-box;
  border: 2px solid #ccc;
  border-radius: 4px;
  background-color: #f8f8f8;
  font-size: 16px;
  resize: none;
}

input[type=text] {
  width: 100%;
  padding: 12px 20px;
  margin: 8px 0;
  box-sizing: border-box;
}

.buttonBlue {
  background-color: #008cba;
  border: none;
  color: white;
  padding: 5px 24px;
  text-align: center;
  text-decoration: none;
  font-size: 16px;
  float: right;
  cursor: pointer;
}

.buttonClear {
  background-color: transparent;
  border: none;
  color: #00e;
  padding: 5px 24px;
  text-align: center;
  text-decoration: none;
  font-size: 16px;
  float: right;
  cursor: pointer;
}

button:disabled,
button[disabled] {
  background-color: rgb(3, 53, 70);
  color: #666;
  cursor: not-allowed;
}
