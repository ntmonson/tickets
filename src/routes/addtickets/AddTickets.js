import React from 'react';
import PropTypes from 'prop-types';
import { graphql, compose } from 'react-apollo';
import withStyles from 'isomorphic-style-loader/lib/withStyles';
import createTicket from './createTicket.graphql';
import ticketsQuery from '../../routes/home/tickets.graphql';
import TicketTable from '../../components/TicketTable/TicketTable';
import s from './AddTickets.css';

class AddTickets extends React.Component {
  static propTypes = {
    addTicket: PropTypes.func.isRequired,
  };

  constructor(props) {
    super(props);
    this.state = {
      newTicketTopic: '',
      newTicketPost: '',
      newTickets: [],
    };
  }

  componentDidMount() {
    this.topicInput.focus();
  }

  handleTopicChange = e => {
    const val = e.target.value;
    this.setState({ newTicketTopic: val });
  };

  handlePostChange = e => {
    const val = e.target.value;
    this.setState({ newTicketPost: val });
  };

  handleAddTicketClick = async () => {
    const { newTicketTopic, newTicketPost } = this.state;
    const data = await this.props.addTicket(newTicketTopic, newTicketPost);
    const { data: { databaseCreateTicket: newTicket } } = data;
    this.setState({ newTickets: [...this.state.newTickets, newTicket] });
    this.clearInputState();
  };

  clearInputState = () => {
    this.setState({ newTicketPost: '', newTicketTopic: '' });
  };

  render() {
    return (
      <div className={s.root}>
        <div className={s.container}>
          <h3>New Ticket:</h3>
          <label htmlFor="topic">
            Topic
            <input
              type="text"
              maxLength="255"
              name="topic"
              ref={input => {
                this.topicInput = input;
              }}
              value={this.state.newTicketTopic}
              onChange={this.handleTopicChange}
            />
          </label>
          <div>
            <label htmlFor="topic">
              Content
              <textarea
                type="textarea"
                maxLength="255"
                name="topic"
                value={this.state.newTicketPost}
                onChange={this.handlePostChange}
              />
            </label>
          </div>
          <button
            disabled={!this.state.newTicketTopic.length}
            className={s.buttonBlue}
            onClick={this.handleAddTicketClick}
          >
            Submit
          </button>
          <button className={s.buttonClear} onClick={this.clearInputState}>
            Clear
          </button>
          {this.state.newTickets.length ? (
            <TicketTable
              tickets={this.state.newTickets}
              showPinnedAndStatus={false}
            />
          ) : (
            <h3>Add new tickets to view them here...</h3>
          )}
        </div>
      </div>
    );
  }
}

const graphqlQueries = compose(
  graphql(createTicket, {
    name: 'addTicket',
    props: ({ addTicket }) => ({
      addTicket: (topic, post) => addTicket({ variables: { topic, post } }),
    }),
    options: {
      update: (proxy, { data: { databaseCreateTicket } }) => {
        const data = proxy.readQuery({ query: ticketsQuery });
        data.databaseGetAllTickets.push(databaseCreateTicket);
        proxy.writeQuery({ query: ticketsQuery, data });
      },
    },
  }),
);

export default compose(withStyles(s), graphqlQueries)(AddTickets);
