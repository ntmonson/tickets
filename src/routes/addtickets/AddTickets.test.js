/* eslint-env jest */
/* eslint-disable padded-blocks, no-unused-expressions */

import React from 'react';
import renderer from 'react-test-renderer';
import '@babel/polyfill';
import { AddTickets } from './AddTickets';

// This oddity is required if the component being tested uses a ref to set focus.
// See: https://reactjs.org/blog/2016/11/16/react-v15.4.0.html
function createNodeMock(element) {
  if (element.type === 'input') {
    return {
      focus() {},
    };
  }
  return null;
}

describe('Add tickets', () => {
  test('AddTickets component renders correctly', () => {
    const component = renderer
      .create(<AddTickets addTicket={() => {}} />, {
        createNodeMock,
      })
      .toJSON();

    expect(component).toMatchSnapshot();
  });
});
