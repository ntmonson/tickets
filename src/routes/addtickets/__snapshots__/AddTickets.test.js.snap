// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add tickets AddTickets component renders correctly 1`] = `
<div
  className="root"
>
  <div
    className="container"
  >
    <h3>
      New Ticket:
    </h3>
    <label
      htmlFor="topic"
    >
      Topic
      <input
        maxLength="255"
        name="topic"
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
    <div>
      <label
        htmlFor="topic"
      >
        Content
        <textarea
          maxLength="255"
          name="topic"
          onChange={[Function]}
          type="textarea"
          value=""
        />
      </label>
    </div>
    <button
      className="buttonBlue"
      disabled={true}
      onClick={[Function]}
    >
      Submit
    </button>
    <button
      className="buttonClear"
      onClick={[Function]}
    >
      Clear
    </button>
    <h3>
      Add new tickets to view them here...
    </h3>
  </div>
</div>
`;
